### PR TITLE
[WIP] rework transformations

### PR DESCRIPTION
**Pull request by AlekseySh**

### CODE REVIEW

**Comment by AlekseySh** [comment]
**CHANGELOG:**
- Augs configs were unified and now they look like other configs & registry
- All of the logic about image transforms moved from Dataset to Transforms (except cropping). So, we don't have im_size and pad_ratio in the dataset anymore
- We don't concatenate norm & to_tensor with user's transforms. All of the augs have to be passed explicitly now
- Dataset will not return any boxes in batch if there were no x1,x2,y1,y2 labels in the original data frame 
- Removed weak albu transforms & ImListDataset
- Reworked logic about im_reader and transforms. Now we select the proper image reader for the given transforms instead of the casting type inside the dataset (pil <-> NumPy) which improved performance & overall logic
- Shuffle=True when we have no sampler
- Added more tests with different mock configs (w boxes, w/o boxes, w categories)


- Due to changed boxes in the dataset Visualizer is broken now. It must be fixed in the next PRs (@b0nce is already working on it)

File: README.md [hunk 1]
```diff
@@ -302,7 +302,7 @@ trainer.fit(pl_model, train_dataloaders=train_loader, val_dataloaders=val_loader
 
 
 ## Models zoo
-|                            model                            | cmc1  |         dataset          |                                      weights & config                                        | hash (the beginning) |
+|                            model                            | cmc1  |         dataset          |                                           weights                                            | hash (the beginning) |
 |:-----------------------------------------------------------:|:-----:|:------------------------:|:--------------------------------------------------------------------------------------------:|:--------------------:|
 | `VitExtractor(weights="vits16_inshop", arch="vits16", ...)` | 0.925 |    DeepFashion Inshop    | [link](https://drive.google.com/drive/folders/1vypEph09rSwKD7iydI4YYZqwZLrdVJPW?usp=sharing) |        384ead        |
 |  `VitExtractor(weights="vits16_sop", arch="vits16", ...)`   | 0.830 | Stanford Online Products | [link](https://drive.google.com/drive/folders/1WfPqCKbZ2KjRRQURGOOwrlQ87EUb7Zra?usp=sharing)   |        85cfa5        |
```
Comment (b0nce) [inline README.md:308]: Maybe fix alignment in the last column

File: examples/visualization.ipynb [hunk 2]
```diff
@@ -73,7 +77,7 @@
     "trainer, ret_dict = pl_val(yaml.load(cfg, Loader=yaml.Loader));\n",
     "\n",
     "clb_metric = [x for x in trainer.callbacks if isinstance(x, MetricValCallback)][0]\n",
-    "vis = RetrievalVisualizer.from_embeddings_metric(clb_metric.metric)\n"
+    "# vis = RetrievalVisualizer.from_embeddings_metric(clb_metric.metric)  # todo: fix in next pr\n"
```
Comment (b0nce) [inline examples/visualization.ipynb:80]: Remove todo?

Comment (b0nce) [reply examples/visualization.ipynb:80]: I mean, if we will fix it later why do we need this todo in master repo at all?

File: examples/visualization.ipynb [hunk 3]
```diff
@@ -83,10 +87,12 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "n_query = clb_metric.metric.distance_matrix.shape[0]\n",
+    "# fix in next pr\n",
```
Comment (b0nce) [inline examples/visualization.ipynb:90]: Same here

File: oml/lightning/entrypoints/train.py [hunk 4]
```diff
@@ -96,11 +95,16 @@ def pl_train(cfg: TCfg) -> None:
         num_workers=cfg["num_workers"],
         batch_size=sampler.batch_size,
         drop_last=True,
+        shuffle=not bool(sampler),
```
Comment (b0nce) [inline oml/lightning/entrypoints/train.py:98]: not sampler will probably also work

File: tests/test_examples/test_runs_via_configs/test_via_configs.py [hunk 5]
```diff
@@ -27,7 +27,9 @@ def run(file: str) -> None:
 
 
 def test_train() -> None:
```
Comment (b0nce) [inline tests/test_examples/test_runs_via_configs/test_via_configs.py:29]: In the future we definitely need something to show fail reasons on this `run(...)` tests
Because now it looks like "Exit with code 1" and no additional info provided

File: tests/test_oml/test_transforms/test_image_augs.py [hunk 6]
```diff
@@ -1,26 +1,31 @@
...
+
+    f_imread = get_im_reader_for_transforms(transforms)
 
     df = pd.read_csv(MOCK_DATASET_PATH / "df.csv")
     df["path"] = df["path"].apply(Path)
 
-    dataset = DatasetWithLabels(
-        df=df, im_size=32, pad_ratio=0, dataset_root=MOCK_DATASET_PATH, transform=transforms, f_imread=f_imread
-    )
+    dataset = DatasetWithLabels(df=df, dataset_root=MOCK_DATASET_PATH, transform=transforms, f_imread=f_imread)
```
Comment (b0nce) [inline tests/test_oml/test_transforms/test_image_augs.py:28]: Additional check on `transforms=None`?

File: oml/datasets/retrieval.py [hunk 7]
```diff
@@ -96,61 +85,59 @@ def _read_bytes_image(path: Union[Path, str]) -> bytes:
...
+
+        elif isinstance(self.transform, torchvision.transforms.Compose):
+            img = img.crop((x1, y1, x2, y2))
+            image_tensor = self.transform(img)
 
-        if pd.isna(row.x_1):
-            x1, y1, x2, y2 = float("nan"), float("nan"), float("nan"), float("nan")
         else:
-            x1, y1, x2, y2 = int(row.x_1), int(row.y_1), int(row.x_2), int(row.y_2)
+            image_tensor = self.transform(img)
```
[the remark this answers is not in the document]
Comment (b0nce) [reply oml/datasets/retrieval.py:109]: Because we now have `ToTensor` as a transform

File: README.md (hunk 1 above)
Comment (AlekseySh) [reply README.md:308]: done

File: examples/visualization.ipynb (hunk 2 above)
Comment (AlekseySh) [reply examples/visualization.ipynb:80]: I prefer to keep it as a reminder

File: oml/lightning/entrypoints/train.py (hunk 4 above)
Comment (AlekseySh) [reply oml/lightning/entrypoints/train.py:98]: using `bool` makes code more readable

File: examples/visualization.ipynb (hunk 3 above)
Comment (AlekseySh) [reply examples/visualization.ipynb:90]: same, prefer to keep as a reminder

File: tests/test_examples/test_runs_via_configs/test_via_configs.py (hunk 5 above)
Comment (AlekseySh) [reply tests/test_examples/test_runs_via_configs/test_via_configs.py:29]: agree

File: tests/test_oml/test_transforms/test_image_augs.py (hunk 6 above)
Comment (AlekseySh) [reply tests/test_oml/test_transforms/test_image_augs.py:28]: done

File: oml/datasets/retrieval.py [hunk 8]
```diff
@@ -96,61 +85,59 @@ def _read_bytes_image(path: Union[Path, str]) -> bytes:
...
 
+        if (not self.bboxes_exist) or any(pd.isna(coord) for coord in [row.x_1, row.x_2, row.y_1, row.y_2]):
+            x1, y1, x2, y2 = 0, 0, im_w, im_h
         else:
-            image_tensor = self.transform(crop)
+            x1, y1, x2, y2 = int(row.x_1), int(row.y_1), int(row.x_2), int(row.y_2)
 
-        row = self.df.iloc[idx]
+        if isinstance(self.transform, albu.Compose):
+            img = img[y1:y2, x1:x2, :]  # todo: since albu may handle bboxes with should move it to augs
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/retrieval.py:101]: thx

File: oml/utils/images/images.py [hunk 9]
```diff
@@ -46,4 +48,25 @@ def imread_pillow(im_src: Union[Path, str, bytes]) -> TImage:
...
 
-__all__ = ["TImReader", "tensor_to_numpy_image", "imread_cv2", "imread_pillow"]
+def pil_from_np_if_needed(img: TImage) -> TPILImage:
+    if isinstance(img, np.ndarray):
+        img = Image.fromarray(img)
+
+    return img
+
+
+def np_from_pil_if_needed(img: TImage) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/utils/images/images.py]: removed

File: tests/test_examples/test_runs_via_configs/test_via_configs.py [hunk 10]
```diff
@@ -27,7 +27,9 @@ def run(file: str) -> None:
...
-    run("train_mock.py")
+    run("train.py")
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply tests/test_examples/test_runs_via_configs/test_via_configs.py:30]: done

File: tests/test_oml/test_transforms/test_image_augs.py [hunk 11]
```diff
@@ -1,26 +1,31 @@
...
-@pytest.mark.parametrize("aug_name", [None, *list(AUGS_REGISTRY.keys())])
-@pytest.mark.parametrize("f_imread", [imread_cv2, imread_pillow])
-def test_transforms(aug_name: Optional[str], f_imread: Any) -> None:
-    transforms = get_augs(name=aug_name) if aug_name is not None else None
+@pytest.mark.parametrize("aug_name", list(TRANSFORMS_REGISTRY.keys()))
+def test_transforms(aug_name: Optional[str]) -> None:
+    path_to_cfg = CONFIGS_PATH / "transforms" / f"{aug_name}.yaml"
+    print(path_to_cfg)
+    with open(path_to_cfg, "r") as f:
+        cfg = OmegaConf.load(f)
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply tests/test_oml/test_transforms/test_image_augs.py:19]: done

File: oml/utils/dataframe_format.py [hunk 12]
```diff
@@ -12,7 +12,7 @@
 
 
 def check_retrieval_dataframe_format(
-    df: Union[Path, str, pd.DataFrame], dataset_root: Optional[Path] = None, sep: str = ","
+    df: Union[Path, str, pd.DataFrame], dataset_root: Optional[Path] = None, sep: str = ",", verbose: bool = False
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/utils/dataframe_format.py:15]: done

File: oml/utils/images/images.py [hunk 13]
```diff
@@ -46,4 +48,25 @@ def imread_pillow(im_src: Union[Path, str, bytes]) -> TImage:
     return image.convert("RGB")
 
 
-__all__ = ["TImReader", "tensor_to_numpy_image", "imread_cv2", "imread_pillow"]
+def pil_from_np_if_needed(img: TImage) -> TPILImage:
+    if isinstance(img, np.ndarray):
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/utils/images/images.py]: removed function

File: oml/datasets/retrieval.py [hunk 14]
```diff
@@ -253,21 +233,19 @@ def get_retrieval_datasets(
     train_dataset = DatasetWithLabels(
         df=df_train,
         dataset_root=dataset_root,
-        im_size=im_size_train,
-        pad_ratio=pad_ratio_train,
-        transform=train_transform,
+        transform=transforms_train,
         cache_size=cache_size,
+        f_imread=get_im_reader_for_transforms(transforms_train),
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/retrieval.py:238]: done

File: oml/transforms/images/albumentations/transforms.py [hunk 15]
```diff
@@ -74,15 +78,34 @@ def get_default_albu() -> albu.Compose:
     """
     augs = albu.Compose(
         [
+            albu.LongestMaxSize(max_size=im_size),
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/transforms/images/albumentations/transforms.py:81]: the opposite: I removed that function

File: oml/datasets/triplet.py [hunk 16]
```diff
@@ -36,27 +35,25 @@ def __init__(
...
 
         self.im_root = im_root
-        self.transforms = transforms
-        self.image_size = image_size
-        self.pad_ratio = pad_ratio
         self.expand_ratio = expand_ratio
         self.f_imread = f_imread
 
+        self.transforms = transforms or get_normalisation_albu()
+        assert isinstance(transforms, albu.Compose) or (transforms is None)
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/triplet.py:54]: because it's only for you now)

File: oml/datasets/retrieval.py [hunk 17]
```diff
@@ -96,61 +85,59 @@ def _read_bytes_image(path: Union[Path, str]) -> bytes:
...
+        im_h, im_w = img.shape[:2] if isinstance(img, np.ndarray) else img.size[::-1]
 
+        if (not self.bboxes_exist) or any(pd.isna(coord) for coord in [row.x_1, row.x_2, row.y_1, row.y_2]):
+            x1, y1, x2, y2 = 0, 0, im_w, im_h
         else:
-            image_tensor = self.transform(crop)
+            x1, y1, x2, y2 = int(row.x_1), int(row.y_1), int(row.x_2), int(row.y_2)
 
-        row = self.df.iloc[idx]
+        if isinstance(self.transform, albu.Compose):
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/retrieval.py:100]: we will force only Compose as augs

File: oml/utils/images/images.py [hunk 18]
```diff
@@ -46,4 +48,9 @@ def imread_pillow(im_src: Union[Path, str, bytes]) -> TImage:
     return image.convert("RGB")
 
 
-__all__ = ["TImReader", "tensor_to_numpy_image", "imread_cv2", "imread_pillow"]
+__all__ = [
```
Comment (b0nce) [inline oml/utils/images/images.py:51]: TImage

Comment (AlekseySh) [reply oml/utils/images/images.py:51]: thx

**Review by b0nce — APPROVED** [review]

**Review by DaloroAT — APPROVED** [review]
